src/irmin-pack/unix/append_only_file_intf.ml

```diff
@@ -29,20 +29,32 @@ module type S = sig
 
   type t
 
+  type flush_callback =
+    [ `Auto of (unit, Io.write_error) result -> unit | `Manual of t -> unit ]
+  (** [flush_callback] is the callback when a flush threshold has been reached.
+
+      - Use [`Auto] callbacks if you only care about the result of a flush. The
+        buffer is flushed automatically before calling the callback with the
+        result.
+      - Use [`Manual] callbacks if you need more control over when the flush
+        occurs, for instance if you want to coordinate with flushing other
+        files. The buffer is not automatically flushed before calling the
+        callback but flush is expected to be called by the callback. *)
```

Comment:
Do we really need a callback for `Auto`? Couldn't the error just be raised by `append_only.ml`?

Reply:
The original motivation for this PR was my previous comment to not require flush to be called in "auto" flush callbacks. The naming doesn't make sense to me but perhaps it is just a historical remnant (looking at `Io_legacy` and config) that we deal with. I guess we can keep everything manually automatic (automatically manual?). :question: 

This is an attempt to clarify the two ways the flush threshold callback (maybe needs a better name) are used in the code, but happy to close this PR and keep code as-is.

Reply:
I think that solving this conundrum requires reserving the word `auto` to the auto/explicit distinction.

What about this?
```
type auto_flush_procedure = [ `Internal | `External of t -> unit ]
```

Reply:
Or 
```
type auto_flush_procedure = [ `Default | `Custom of t -> unit ]
```

Reply:
I think I lean towards `Internal/External`. `Internal` still needs a function associated if we want to raise excretions on error.

The auto prefix here is still weird/confusing to me, but I think I understand more where you are coming from now. It is more of a `threshold_flush` than an `auto_flush` in my mind but `threshold_flush_procedure` is quite the name :upside_down_face:.

(as an aside, and of no real importance, the typical english distinction is auto/manual or implicit/explicit)

Reply:
There is a misunderstanding somewhere 🤔 

On main, `auto flush` refers to flushes automatically triggered by the file stack code because a write buffer is full while `flush` (no prefix) is a call to `File_manager.flush` at the end of batch.

Also, wouldn't it be possible for `append_only` to call `Errs.raise_if_error` in the Auto aka Internal aka Default case? This was there is no need to pass a callback to raise the error

Reply:
We can make it completely auto/internal/default/magic if we add `Errs` to the functor. 

Re: auto, this seems more of a concern of file manager than append only but I'm fine to leave it. :ballot_box_with_check:

Reply:
Took another pass at this based on our discussion.